### PR TITLE
Move to py3.10, django 3.2 and heroku stack 22

### PR DESCRIPTION
**Pull request by gedankenstuecke**
## Description

Okay, this started out as the wish to be able to have a recaptcha when registering to make live for our spammers at least a bit harder, but then it turned out to be a bit more work given that our deployment was a bit behind in terms of heroku stack. 

So in short this one: 
1. Updates our heroku stack to 22
2. Bumps the Python version from 3.6 to 3.10 (as that is supported on Heroku stack 22)
3. Bumps the Django library to 3.2 (from 2.2) (and many downstream libraries), as those are needed for the newer Python version. 
4. Finally adds in the captcha, which started this journey

## Testing

I've mainly tested it on our staging deployment on Heroku, which I bumped to Stack22 already, the deployment there seems to work as far as I could see.

### CODE REVIEW

File: data_import/views.py [hunk 1]
```diff
@@ -59,8 +59,10 @@ def get_and_log(self, request, key_object=None):
 
         user = request.user if request.user.is_authenticated else None
 
+        ip_address, _ = get_client_ip(request)
```
Comment (gedankenstuecke) [inline data_import/views.py:62]: The new `get_client_ip` function returns two values instead of one.

File: open_humans/forms.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-from captcha.fields import ReCaptchaField
+from django_recaptcha.fields import ReCaptchaField
```
Comment (gedankenstuecke) [inline open_humans/forms.py:1]: the update in recaptcha renamed the library

File: open_humans/templatetags/utilities.py [hunk 3]
```diff
@@ -95,7 +95,7 @@ def markdown(value):
     """
     cleaned = bleach.clean(
         markdown_library.markdown(value),
-        tags=bleach.ALLOWED_TAGS + ["p", "h1", "h2", "h3", "h4", "h5", "h6"],
+        tags=list(bleach.ALLOWED_TAGS) + ["p", "h1", "h2", "h3", "h4", "h5", "h6"],
```
Comment (gedankenstuecke) [inline open_humans/templatetags/utilities.py:98]: Newer python (or lib) doesn't like mixing types, so need to convert the former.

File: public_data/models.py [hunk 4]
```diff
@@ -34,10 +34,6 @@ def public_count(project):
             project_membership__project=project,
             # Filter to only count members with datafiles for this project.
             is_public=True,
-            project_membership__project__in=F(
```
Comment (gedankenstuecke) [inline public_data/models.py:37]: For some reason this kept failing, as somewhere along the way the way these interpreted are changed. Before the SQL query ended up being this

```
SELECT "public_data_publicdataaccess"."id", "public_data_publicdataaccess"."participant_id", "public_data_publicdataaccess"."project_membership_id", "public_data_publicdataaccess"."is_public" FROM "public_data_publicdataaccess" INNER JOIN "private_sharing_datarequestprojectmember" ON ("public_data_publicdataaccess"."project_membership_id" = "private_sharing_datarequestprojectmember"."id") INNER JOIN "open_humans_member" ON ("private_sharing_datarequestprojectmember"."member_id" = "open_humans_member"."id") INNER JOIN "auth_user" ON ("open_humans_member"."user_id" = "auth_user"."id") INNER JOIN "data_import_datafile" ON ("auth_user"."id" = "data_import_datafile"."user_id") INNER JOIN "private_sharing_projectdatafile" ON ("data_import_datafile"."id" = "private_sharing_projectdatafile"."parent_id") WHERE ("public_data_publicdataaccess"."is_public" = True AND "private_sharing_datarequestprojectmember"."project_id" = 92 AND "private_sharing_datarequestprojectmember"."project_id" IN ("private_sharing_projectdatafile"."direct_sharing_project_id"))
```
(note the parentheses for the last `AND`), with the updates here it ended up being 

```
SELECT "public_data_publicdataaccess"."id", "public_data_publicdataaccess"."participant_id", "public_data_publicdataaccess"."project_membership_id", "public_data_publicdataaccess"."is_public" FROM "public_data_publicdataaccess" INNER JOIN "private_sharing_datarequestprojectmember" ON ("public_data_publicdataaccess"."project_membership_id" = "private_sharing_datarequestprojectmember"."id") INNER JOIN "open_humans_member" ON ("private_sharing_datarequestprojectmember"."member_id" = "open_humans_member"."id") INNER JOIN "auth_user" ON ("open_humans_member"."user_id" = "auth_user"."id") INNER JOIN "data_import_datafile" ON ("auth_user"."id" = "data_import_datafile"."user_id") INNER JOIN "private_sharing_projectdatafile" ON ("data_import_datafile"."id" = "private_sharing_projectdatafile"."parent_id") WHERE ("public_data_publicdataaccess"."is_public" AND "private_sharing_datarequestprojectmember"."project_id" = 92 AND "private_sharing_datarequestprojectmember"."project_id" IN "private_sharing_projectdatafile"."direct_sharing_project_id")
```

Given that the only effect is "miscounting" the number of users (by including those that do have no files but are willing to make them public in principle) I think that's acceptable

**Comment by gedankenstuecke** [comment]
@beaugunderson I've tagged you in, as I think you'd be most likely in the position to give this a look to see if that all makes sense or not. You'll also have access to the heroku stacks to see everything there I think :-)

**Review by beaugunderson — APPROVED** [review]
everything looks like about what I'd expect :)

**Comment by gedankenstuecke** [comment]
Thanks @beaugunderson! Did you have a chance to run the tests locally? I didn't yet myself but can try to sort that out over the WE hopefully!

**Comment by beaugunderson** [comment]
> Thanks @beaugunderson! Did you have a chance to run the tests locally? I didn't yet myself but can try to sort that out over the WE hopefully!

there are failures:

```
======================================================================
FAIL: test_authorize_if_already_authorized (private_sharing.tests.DirectSharingOAuth2Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/beau/p/open-humans/private_sharing/tests.py", line 247, in test_authorize_if_already_authorized
    self.assertTrue(b"Project previously authorized." in response.content)
AssertionError: False is not true

======================================================================
FAIL: test_authorize_if_logged_in (private_sharing.tests.DirectSharingOAuth2Tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/beau/p/open-humans/private_sharing/tests.py", line 238, in test_authorize_if_logged_in
    self.assertEqual(response.status_code, 200)
AssertionError: 302 != 200

----------------------------------------------------------------------
Ran 76 tests in 13.755s

FAILED (failures=2, errors=5, skipped=7)
```

Additionally, I did not have AMQP running so both `test_message_member` tests hung, I commented them out to see if the rest succeeded.

**Comment by gedankenstuecke** [comment]
Thanks, I just tried reproducing both of those tests in the staging environment and the joining for the on-site worked, as did the "previously joined" warning!

**Pull request merged**